File: 2.两数相加.cpp
/*
 * @lc app=leetcode.cn id=2 lang=cpp
 *
 * [2] 两数相加
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode *ret=new ListNode,*tmp=new ListNode;
        tmp->next=ret;
        int s=0, now_val;
        while(l1!=nullptr&&l2!=nullptr){
            tmp=tmp->next;
            now_val=l1->val+l2->val+s;
            tmp->val=now_val%10;
            s=now_val/10;
            l1=l1->next;
            l2=l2->next;
            tmp->next=new ListNode;
        }
        while(l1!=nullptr){
            tmp=tmp->next;
            now_val=l1->val+s;
            tmp->val=now_val%10;
            s=now_val/10;
            if(s==0){
                tmp->next=l1->next;
                return ret;
            }
            tmp->next=new ListNode;
            l1=l1->next;
        }
        while(l2!=nullptr){
            tmp=tmp->next;
            now_val=l2->val+s;
            tmp->val=now_val%10;
            s=now_val/10;
            if(s==0){
                tmp->next=l2->next;
                return ret;
            }
            tmp->next=new ListNode;
            l2=l2->next;
        }
        if(s)tmp->next->val=s;
        else tmp->next=nullptr;
        return ret;
    }
};
// @lc code=end

File: 2.两数相加.py
#
# @lc app=leetcode.cn id=2 lang=python3
#
# [2] 两数相加
#

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# @lc code=start
from typing import List, Optional
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def func(self, l1: Optional[ListNode], l2: Optional[ListNode], carry: int = 0):
        if l1 is not None and l2 is not None:
            return ListNode((l1.val + l2.val + carry) % 10, self.func(l1.next, l2.next, (l1.val + l2.val + carry) // 10))
        if l1 is l2 is None:
            if carry == 0: return None
            return ListNode(carry % 10, self.func(None, None, carry // 10))
        if l1 is None: l1, l2 = l2, l1
        return ListNode((l1.val + carry) % 10, self.func(l1.next, None, (l1.val + carry) // 10))
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        return self.func(l1, l2)
# @lc code=end

